### PR TITLE
Dynamically type getters and hassers for parameters

### PR DESCRIPTION
**Pull request by Kocal**
Hi!
This is a proposal for #127.

I've re-used the same logic for services:
- read and parse the Symfony container XML dump (it would have been simpler with a PHP dump 😛)
- create a `ParameterMap` which contains all parameters

Dynamic return types have been implemented for:
- `ParameterBagInterface#get` and `ParameterBagInterface#has`
- `ContainerInterface#getParameter` and `ContainerInterface#hasParameter` 
- `AbstractController#getParameter` and `Controller#getParameter` 

~However, I'm not able to do the same for `AbstractController#getParameter` and `Controller#getParameter` when using `$this->getParameter('...')` inside a `AbstractController` and `Controller`. It always use the [return type from `AbstractController#getParameter`](https://github.com/symfony/symfony/blob/5.x/src/Symfony/Bundle/FrameworkBundle/Controller/AbstractController.php#L78). Is this a limitation from PHPStan or I did something wrong @ondrejmirtes?~

~Also in Symfony 4.0 (lowest deps), there is no method `AbstractController#getParameter()` and it make the tests failing. What is the best way to support lowest and highest Symfony version in tests and extension? What about using Symfony 4.4 as minimal requirements and dropping `Controller` usage? This is the only 4.x supported version right now.~
**EDIT:** Nevermind I fixed in c2d3f798e84fc9949f60ff28f45f09dc9d6cebe4 by using 4.4 as lowest version for `symfony/framework-bundle`, and manually create the default type if a parameter does not exist.

### CODE REVIEW

File: tests/Type/Symfony/ExtensionTestWithoutContainer.php [hunk 1]
```diff
@@ -10,7 +10,15 @@ class ExtensionTestWithoutContainer extends TypeInferenceTestCase
...
-		yield from $this->gatherAssertTypes(__DIR__ . '/data/ExampleControllerWithoutContainer.php');
+		if (class_exists('Symfony\Bundle\FrameworkBundle\Controller\Controller')) {
+			yield from $this->gatherAssertTypes(__DIR__ . '/data/ExampleController.php');
+		}
+
+		if (!class_exists('Symfony\Bundle\FrameworkBundle\Controller\AbstractController')) {
+			return;
+		}
+
+		yield from $this->gatherAssertTypes(__DIR__ . '/data/ExampleAbstractController.php');
```
Comment (ondrejmirtes) [inline tests/Type/Symfony/ExtensionTestWithoutContainer.php:21]: I know that PHPCS forced you to do this but I'd prefer structure that allowed adding new files easily. So I'd suggest adding another dataProvider with a single `class_exists` condition and `yield from` with just a single file. And adding that provider to `testFileAsserts`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Hi, I love this! This is well thought out and I found only a single minor problem :) I wish that more people like you contributed to PHPStan :) Thank you.

File: tests/Type/Symfony/ExtensionTestWithoutContainer.php (hunk 1 above)
Comment (ondrejmirtes) [reply tests/Type/Symfony/ExtensionTestWithoutContainer.php:21]: Actually I might resolve this myself because I really want to get this released 👍

**Comment by ondrejmirtes** [comment]
Alright: https://github.com/phpstan/phpstan-symfony/pull/152/commits/a586e2495c2622d5494999e85c77ddd5573efaad

I'm gonna squash this and release it :)

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by mhujer** [comment]
@Kocal 👍 👏  Just tested your branch on our project and it found the following:
(all of them are valid reports which need to be fixed in our code)

- `Offset 0 does not exist on array<string>|string|null.`
- `Cannot cast array<string>|string|null to int.` (at `(int) $rawAmount`)
- ` Parameter #1 $input of method FooBar::fromString() expects string|null, array<string>|string|null given.` (when passing a `$request->query->all()['amount']`)
- `Parameter #1 $parameters of method BarBaz::filterParameters() expects array<string>, array<string, array<string>|string> given` when passing `$request->query->all()`

**Comment by ondrejmirtes** [comment]
Enjoy it in phpstan-symfony 0.12.29 :)

**Comment by Kocal** [comment]
Nice! :D

**Pull request merged**